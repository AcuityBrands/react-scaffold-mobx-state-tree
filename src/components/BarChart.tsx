/**
 * Bar Chart Component
 * 
 * Renders input dataset as basic bar chart
 */

import * as React from "react";
import * as Chart from 'chart.js/dist/Chart.js'

interface IBarChartProps {
  id: string,
  label: string,
  height: number,
  data: IChartData
}

interface IChartData {
  labels: Array<string>,
  values: Array<number>
}

export class BarChart extends React.Component<IBarChartProps, null> {
  
  componentDidMount() {
    this.renderChart();
  }

  renderChart() {
    //Chart globals
    Chart.defaults.global.legend.position = "bottom"

    // TODO: Figure out why chart data won't use 
    // props without a clone of the array

    //Initialize Chart
    new Chart(this.props.id, {
      type: 'bar',
      data: {
        labels: this.props.data.labels,
        datasets: [{
          label: this.props.label,
          data: this.props.data.values.slice(),
          backgroundColor: this.getRandomColor(),
          pointRadius: 0,
          borderColor: 'rgba(0,0,0,0)'
        }]
      },
      options: {
        legend: { display: false },
        scales: {
          xAxes: [{
              gridLines: {
                  display: false
              }
          }],
          yAxes: [{
            gridLines: {
                borderDash: [2,2]
            }
          }]
        },
        tooltips: {
          xPadding: 10,
          yPadding: 10,
          titleMarginBottom: 8
        }
      }
    });
  }

  // Create canvas to render chart
  render() {
    return (
      <canvas id={this.props.id} height={this.props.height}></canvas>
    );
  }

  // Randomizes the color of the chart - hardcoded opacity level
  getRandomColor() {
    const o = Math.round, r = Math.random, s = 255;
    return 'rgba(' + o(r()*s) + ',' + o(r()*s) + ',' + o(r()*s) + ',0.7)';
  }
}