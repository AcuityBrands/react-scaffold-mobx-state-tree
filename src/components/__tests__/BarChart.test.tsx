import * as React from "react";
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { BarChart } from '../BarChart';

Enzyme.configure({ adapter: new Adapter() });

const setup = propOverrides => {
  let spy
  const props = Object.assign({
    id: 'hello',
    label: 'test',
    height: 50,
    data: {
      labels: ['label'],
      values: [1]
    }
  }, propOverrides)
  const wrapper = Enzyme.shallow(<BarChart {...props} />)
  return {
    props,
    wrapper
  }
}

describe('BarChart Component', () => {
  beforeEach(() => {
    // Mock constructing the chart since we don't have full access to 
    // a canvas element with Jest.
    BarChart.prototype.renderChart = jest.fn();
  });

  it("will create a canvas element for the chart", () => {
    const { wrapper, props } = setup({});
    expect(wrapper.find('canvas').length).toBe(1);
  })

  it("matches the snapshot", () => {
    const { wrapper } = setup({});
    expect(wrapper).toMatchSnapshot();
  })

  it("will create and render the chart config on startup", () => {
    const spy = jest.spyOn(BarChart.prototype, "renderChart")
    const { wrapper, props } = setup({});
    expect(spy).toHaveBeenCalledTimes(1);
    spy.mockClear();
  })
})