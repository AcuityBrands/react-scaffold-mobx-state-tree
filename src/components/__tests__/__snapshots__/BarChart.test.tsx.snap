// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart Component matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BarChart
    data={
      Object {
        "labels": Array [
          "label",
        ],
        "values": Array [
          1,
        ],
      }
    }
    height={50}
    id="hello"
    label="test"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "height": 50,
      "id": "hello",
    },
    "ref": null,
    "rendered": null,
    "type": "canvas",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "height": 50,
        "id": "hello",
      },
      "ref": null,
      "rendered": null,
      "type": "canvas",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
